plugins/ndpi/ndpi.c

```diff
@@ -0,0 +1,624 @@
+}
+
+static void NdpInitRiskKeyword(void)
+{
+    /* SCSigTableElmt and DetectHelperKeywordRegister don't yet
+     * support all the fields required to register the nDPI keywords,
+     * so we'll just register with an empty keyword specifier to get
+     * the ID, then fill in the ID. */
+    SCSigTableElmt keyword = {};
+    ndpi_protocol_keyword_id = DetectHelperKeywordRegister(&keyword);
+    SCLogDebug("Registered new ndpi-protocol keyword with ID %" PRIu32, ndpi_protocol_keyword_id);
+
+    sigmatch_table[ndpi_protocol_keyword_id].name = "ndpi-protocol";
+    sigmatch_table[ndpi_protocol_keyword_id].desc = "match on the detected nDPI protocol";
+    sigmatch_table[ndpi_protocol_keyword_id].url = "/rules/ndpi-protocol.html";
+    sigmatch_table[ndpi_protocol_keyword_id].Match = DetectnDPIProtocolPacketMatch;
+    sigmatch_table[ndpi_protocol_keyword_id].Setup = DetectnDPIProtocolSetup;
+    sigmatch_table[ndpi_protocol_keyword_id].Free = DetectnDPIProtocolFree;
+    sigmatch_table[ndpi_protocol_keyword_id].flags =
+            (SIGMATCH_QUOTES_OPTIONAL | SIGMATCH_HANDLE_NEGATION);
+
+    ndpi_risk_keyword_id = DetectHelperKeywordRegister(&keyword);
+    SCLogDebug("Registered new ndpi-risk keyword with ID %" PRIu32, ndpi_risk_keyword_id);
+
+    sigmatch_table[ndpi_risk_keyword_id].name = "ndpi-risk";
+    sigmatch_table[ndpi_risk_keyword_id].desc = "match on the detected nDPI risk";
+    sigmatch_table[ndpi_risk_keyword_id].url = "/rules/ndpi-risk.html";
+    sigmatch_table[ndpi_risk_keyword_id].Match = DetectnDPIRiskPacketMatch;
+    sigmatch_table[ndpi_risk_keyword_id].Setup = DetectnDPIRiskSetup;
+    sigmatch_table[ndpi_risk_keyword_id].Free = DetectnDPIRiskFree;
+    sigmatch_table[ndpi_risk_keyword_id].flags =
+            (SIGMATCH_QUOTES_OPTIONAL | SIGMATCH_HANDLE_NEGATION);
+}
+
+static void NdpiInit(void)
+{
+    SCLogDebug("Initializing nDPI plugin");
+
+    /* Register thread storage. */
+    thread_storage_id = ThreadStorageRegister("ndpi", sizeof(void *), NULL, ThreadStorageFree);
+    if (thread_storage_id.id < 0) {
+        FatalError("Failed to register nDPI thread storage");
+    }
+
+    /* Register flow storage. */
+    flow_storage_id = FlowStorageRegister("ndpi", sizeof(void *), NULL, FlowStorageFree);
+    if (flow_storage_id.id < 0) {
+        FatalError("Failed to register nDPI flow storage");
+    }
+
+    /* Register flow lifecycle callbacks. */
+    SCFlowRegisterInitCallback(OnFlowInit, NULL);
+    SCFlowRegisterUpdateCallback(OnFlowUpdate, NULL);
+
+    /* Not needed for nDPI, but exists for completeness. */
+    SCFlowRegisterFinishCallback(OnFlowFinish, NULL);
+
+    /* Register thread init callback. */
+    SCThreadRegisterInitCallback(OnThreadInit, NULL);
+
+    /* Register an EVE callback. */
+    SCEveRegisterCallback(EveCallback, NULL);
+
+    NdpInitRiskKeyword();
+}
+
+const SCPlugin PluginRegistration = {
+    .name = "ndpi",
+    .author = "Luca Deri",
+    .license = "GPLv2",
```

Comment:
@lucaderi @cardigliano GPLv3 right?

Reply:
I guess this comes from a cut&paste. nDPI is LGPLv3.